File: Javascript/1857LargestColorValueInADirectedGraph.js
/**
 * @param {string} colors
 * @param {number[][]} edges
 * @return {number}
 */
var largestPathValue = function(colors, edges) {
    let ans = 0;
    let processed = 0;
    const n = colors.length;
    const inDegrees = new Array(n).fill(0);
    const graph = new Array(n);
    const count = new Array(n);
    
    for (let i = 0; i < n; i++) {
        graph[i] = [];
        count[i] = new Array(26).fill(0);
    }
    
    for (const [u, v] of edges) {
        graph[u].push(v);
        inDegrees[v]++;
    }
    
    const queue = [];
    for (let i = 0; i < n; i++) {
        if (inDegrees[i] === 0) {
            queue.push(i);
        }
    }
    
    while (queue.length > 0) {
        const u = queue.shift();
        processed++;
        ans = Math.max(ans, ++count[u][colors.charCodeAt(u) - 'a'.charCodeAt(0)]);
        
        for (const v of graph[u]) {
            for (let i = 0; i < 26; i++) {
                count[v][i] = Math.max(count[v][i], count[u][i]);
            }
            if (--inDegrees[v] === 0) {
                queue.push(v);
            }
        }
    }
    
    return processed === n ? ans : -1;
};

File: Java/1857LargestColorValueInADirectedGraph.java
class Solution {
    public int largestPathValue(String colors, int[][] edges) {
        int ans = 0;
        int processed = 0;
        int[] inDegrees = new int[colors.length()];
        List<Integer>[] graph = new List[colors.length()];
        int[][] count = new int[colors.length()][26];


        for (int i = 0; i < colors.length(); ++i) {
            graph[i] = new ArrayList<>();
        }

        for(int[] edge: edges){
            final int u = edge[0];
            final int v = edge[1];
            graph[u].add(v);
            ++inDegrees[v];
        }

        Queue<Integer> q = IntStream.range(0, colors.length())
                           .filter(i -> inDegrees[i] == 0)
                           .boxed()
                           .collect(Collectors.toCollection(ArrayDeque::new));

        while(!q.isEmpty()) {
            final int out = q.poll();
            ++processed;
            ans = Math.max(ans, ++count[out][colors.charAt(out) - 'a']);

            for (final int in: graph[out]) {
                for (int i = 0; i < 26; ++i) {
                    count[in][i] = Math.max(count[in][i], count[out][i]);
                }

                if(--inDegrees[in] == 0){
                    q.offer(in);
                }
            }
        }

        return processed == colors.length() ? ans : -1;
    }
}
